keps/sig-api-machinery/20190226-network-proxy.md

```diff
@@ -0,0 +1,618 @@
+on the local network without pointing to any proxies, 
+which the other 2 networks going through either master or worker proxy. 
+Please note that while these are represented as resources 
+they are not intended to be loaded dynamically. 
+They are intended to be provided as configuration at KAS start time.
+
+```yaml
+apiVersion: connectivityservice.k8s.io/v1alpha1
+kind: ConnectivityServiceConfiguration
+connectionService:
+  name: direct
+  connection:
+    type: direct
+---
+apiVersion: connectivityservice.k8s.io/v1alpha1
+kind: ConnectivityServiceConfiguration
+connectionService:
+  name: Master
+  connection:
+    type: grpc
+    url: grpc://1.2.3.4:5678
+    caBundle: file1.ca
+---
+apiVersion: connectivityservice.k8s.io/v1alpha1
+kind: ConnectivityServiceConfiguration
+connectionService:
+  name: ZoneA
+  connection:
+    type: grpc
+    url: grpc://1.2.3.4:5679
+    caBundle: file2.ca
+  matchRule:
+    ruleType: label
+    label: failure-domain.beta.kubernetes.io/zone
+    value: ZoneA
+---
+apiVersion: connectivityservice.k8s.io/v1alpha1
+kind: ConnectivityServiceConfiguration
+connectionService:
+  name: ZoneB
+  connection:
+    type: grpc
+    url: grpc://1.2.3.4:5680
+    caBundle: file3.ca
+  matchRule:
+    ruleType: label
+    label: failure-domain.beta.kubernetes.io/zone
+    value: ZoneB
+
+```
+
+NetworkContext could potentially be extended to contain 
+more contextual information in the future to do smarter routing 
+based on, for instance, which destination k8s object API server 
+tries to connect, which user/tenant tries to initiate the request, etc.
+
+### Proxy gRPC definition
+
+```grpc
+syntax = "proto3";
+
+service ProxyService {
+    // Proxy connects to a remote address by stream id, and establish
+    // a bi-directional stream of packet exchange.
+    rpc Proxy(stream Packet) returns (stream Packet) {}
+}
+
+enum PacketType {
+  DIAL_REQ = 0;
+  DIAL_RSP = 1;
+  CLOSE_REQ = 2;
+  CLOSE_RSP = 3;
+  DATA = 4;
+}
+
+message Packet {
+  PacketType type = 1;
+
+  oneof payload {
+    DialRequest dialRequest = 2;
+    DialResponse dialResponse = 3;
+    CloseRequest closeRequest = 4;
+    CLoseResponse closeResponse = 5;
+    Data data = 6;
+  }
+}
+
+// Error is sent when error happens from remote side of connection when it tries
+// to read or write from it
+enum Error {
+  EOF = 0;
+  EIO = 1;
+  ECONNRESET = 2;
+  ETIMEOUT = 3;
+  EADDRNOTAVAIL = 4;
+  EMFILE = 5;
+  // … 
+}
+
+// DialRequest represents a request to dial to an address on the other 
+// side of the tunnel. The format is inspired by golang's net interface
+// https://golang.org/pkg/net/#Dial 
+message DialRequest {
+    // network representing a named network. "Tcp" is the only supported
+    // value
+    string network = 1;
+
+    // For TCP network, the address has the form "ip:port", where host must
+    // be IP address.
+    string address = 2;
+
+    // random is the randomly generated bytes that represents a given dial request.
```

Comment:
Call it `nonce` maybe - I think that's what it is?

Reply:
+1